### PR TITLE
Add Learning Rate Scheduler from GraphCast

### PR DESCRIPTION
**Pull request by matschreiner**
This PR implements 2 features. 

1) A `--steps` argument has been added to the training script, allowing control over the number of training steps before termination. This is implemented using `Trainer(..., max_num_steps=args.steps, ...)` in pytorch_lightning. If both `max_num_epochs` and `max_num_steps` are specified in the `Trainer`, training will stop when either has been reached. By default `"--steps"` is -1 which means that without specifying this keyword, training will never terminate due to `max_num_steps`. 

2) A learning scheduler to NeuralLAM following the specifications given for learning rates in GraphCast. The defaults are implemented such that when using the default learning rate for Adam (`1e-3`) the linear warmup will range from epsilon to `1e-3` and then cosine annealing from 1e-3 to 1e-6. Otherwise, min and max learning rate will range from `min_factor * initial_lr` to `max_factor * initial_lr)`
The signature and defaults of the scheduler is

```
neural_lam.lr_scheduler.WarmupCosineAnnealingLR(
        optimizer,
        warmup_steps=100000,
        annealing_steps=1000,
        max_factor=1.0,
        min_factor=0.001,
)
```

The scheduler has 3 phases:

## Warmup phase
In this phase the learning rate is warmed up by linearly increasing the learning rate from `min_factor * initial_learning_rate` to `max_factor * initial_learning_rate` (The learning rate is defined in the optimizer).

## Annealing phase
In this phase the learning rate is annealed by using a CosineAnnealing schedule that anneals the learning rate back to `min_factor * initial_learning_rate` by following a half a cosing cycle. 

## Fine tuning phase
In this phase the annealed learning rate is used until the training terminates.

# Learning schedule
Using the default learning rate for the adam optimizer (1e-3), 20 warmup steps, 100 annealing steps and 150 steps in total will produce a schedule that looks as folllows: 

```
WarmupCosineAnnealingLR(
        optimizer,
        warmup_steps=20,
        annealing_steps=100,
)
```

image

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

## Type of change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [x] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [x] I have performed a self-review of my code
- [ ] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [x] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [x] I have updated the [README](README.MD) to cover introduced code changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [x] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [ ] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug

## Checklist for assignee

- [x] PR is up to date with the base branch
- [x] the tests pass
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed* or *fixed*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

File: neural_lam/lr_scheduler.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+        super().__init__(optimizer)
+
+    def get_lr(self):
+        if self._step_count <= self.warmup_steps:
+            return self.warmup_scheduler.get_last_lr()
+        elif self._step_count <= self.warmup_steps + self.annealing_steps:
+            self.annealing_scheduler.step()
+
+        return True
```
Comment (SimonKamuk) [inline neural_lam/lr_scheduler.py:39]: what does it mean for get_lr to return True? 

Why are we taking an annealing step here?

File: tests/test_lr_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    annealing_lr = lrs[warmup_steps : warmup_steps + annealing_steps]
+
+    # Formula for the cosine annealing
+    expected_annealing_lr = min_factor * initial_lr + 0.5 * (
+        max_factor * initial_lr - min_factor * initial_lr
+    ) * (1 + np.cos(np.pi * np.arange(annealing_steps) / annealing_steps))
+    assert np.allclose(annealing_lr, expected_annealing_lr)
+
+    last_lr = lrs[-1]
+    assert all(lrs[-5:] == np.ones(5) * last_lr)
```
Comment (SimonKamuk) [inline tests/test_lr_scheduler.py:58]: This seems to just make sure the last 5 points are constant. Shouldn't this be testing if all final steps have the expected value? something more like
```suggestion
    end_lr = np.array(lrs[warmup_steps + annealing_steps:])
    assert all(end_lr ==  min_factor * initial_lr)
```

**Review by SimonKamuk — CHANGES_REQUESTED** [review]
This is really good, I just have a few minor comments. But also, shouldn't this be implemented in train_model.py or ar_model.py so it can actually be used?

**Comment by matschreiner** [comment]
> This is really good, I just have a few minor comments. But also, shouldn't this be implemented in train_model.py or ar_model.py so it can actually be used?

Thanks :) Yes, this actually doesn't let you use the scheduler! Configuring the scheduler in the framework is another can of worms, and to limit the scope of the PR I only have the implementation of the scheduler here so that we can use it in a later PR.

File: neural_lam/lr_scheduler.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            pass
+        elif self._step_count <= self.warmup_steps:
+            self.warmup_scheduler.step()
+        elif self._step_count <= self.warmup_steps + self.annealing_steps:
+            self.annealing_scheduler.step()
+        self._step_count += 1
+
+
+if __name__ == "__main__":
+    # Run this code to visualize the learning rate schedule
```
[the remark this answers is not in the document]
Comment (matschreiner) [reply neural_lam/lr_scheduler.py:52]: Yes, you are right. :) 
We might also want to visualize other schedulers, so I will remove it and make another PR where I generalise it.

File: neural_lam/lr_scheduler.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        self,
+        optimizer,
+        warmup_steps=1000,
+        annealing_steps=100000,
+        max_factor=1.0,
+        min_factor=0.001,
+    ):
+        self.warmup_steps = warmup_steps
+        self.annealing_steps = annealing_steps
+        initial_learning_rate = optimizer.param_groups[0]["lr"]
```
[the remark this answers is not in the document]
Comment (matschreiner) [reply neural_lam/lr_scheduler.py]: Good suggestion. Thanks, I'm now making sure that we don't have multiple parameter groups and I added a TODO that we should generalize this when we have time.

**Comment by matschreiner** [comment]
@SimonKamuk thanks for your review, I commented on your comments and updated the PR :)

File: neural_lam/lr_scheduler.py (hunk 1 above)
Comment (SimonKamuk) [reply neural_lam/lr_scheduler.py]: Great, thanks for the clarification. Maybe we could add a comment explaining what the return value of True means, then I think we're good to go!

**Comment by SimonKamuk** [comment]
Looking great, one last comment and then the linting errors and this is done I think :)

**Comment by joeloskarsson** [comment]
@matschreiner Is it correct that this is superseded by #107 ?

**Comment by matschreiner** [comment]
> @matschreiner Is it correct that this is superseded by #107 ?

Hi @joeloskarsson! No they are different PRs, as far as I remember - this PR implements the graphcast learning rate scheduler, the other PR makes LR scheduler configurable through the neurallam_config.

**Comment by joeloskarsson** [comment]
Looking at lr scheduling again I guess I am still a bit confused about this and #107. If I wanted to use the graphcast-like lr scheduling, why would I not just configure that in a config like #107? Is the problem that one can't use ChainedScheduler there (https://github.com/mllam/neural-lam/pull/107#issuecomment-3601099135), so this PR would be an alternative to that that directly adds the graphcast-like scheduler as a separate one?

**Comment by observingClouds** [comment]
I agree, the implementation of the `WarmupCosineAnnealingLR` scheduler, can be achieved with the `ChainedScheduler`:

```python
    def configure_optimizers(self):
        opt = torch.optim.AdamW(
            self.parameters(), lr=self.args.lr, betas=(0.9, 0.95)
        )
        scheduler1 = torch.optim.lr_scheduler.LinearLR(
            opt,
            total_iters=10
        )
        scheduler2 = torch.optim.lr_scheduler.CosineAnnealingLR(
            opt,
            T_max=120,
            eta_min=self.args.min_lr if hasattr(self.args, "min_lr") else 0.0,
        )
        scheduler3 = torch.optim.lr_scheduler.ConstantLR(
            opt,
            factor=1,
            total_iters=0,
        )
        scheduler = torch.optim.lr_scheduler.ChainedScheduler([scheduler1,scheduler2,scheduler3], optimizer=opt)
        return {
            "optimizer": opt,
            "lr_scheduler": {
                "scheduler": scheduler,
                "interval": "epoch",
                "frequency": 1,
            },
        }
 ```

**Comment by sadamov** [comment]
I make this a draft PR for now, because according to the discussion its main feature is already covered by https://github.com/mllam/neural-lam/pull/107

**Comment by sadamov** [comment]
Closing this one. Per the design discussion with @joeloskarsson and @observingClouds, GraphCast's warmup + cosine annealing can be achieved via `ChainedScheduler(LinearLR, CosineAnnealingLR)` in #107 without needing a dedicated `WarmupCosineAnnealingLR` class. Closing in favour of consolidating the LR scheduler work in #107. Thanks @matschreiner.
